### PR TITLE
Correctly decode negative timestamps with Avro

### PR DESCRIPTION
**Pull request by umanwizard**
Our algorithm was just totally wrong for negative numbers; copy the algorithm Chrono uses instead. 

### Motivation

  * This PR fixes a previously unreported bug.

Reported on Slack [here](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1670889931262329). Note that I'm not sure whether there are other places in Materialize where negative timestamps break things, but this at least fixes the issue for the Avro decoder.

### Tips for reviewer

See the `// TODO` comment in the codebase. I've filed an issue with Chrono here: https://github.com/chronotope/chrono/issues/903

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Fix incorrect decoding of pre-1970 timestamps in Avro records

### CODE REVIEW

**Review by petrosagg — APPROVED** [review]

**Review by aljoscha — APPROVED** [review]

File: src/avro/src/decode.rs [hunk 1]
```diff
@@ -110,23 +110,116 @@ impl Display for TsUnit {
...
-    ))
+    let nanos_per_unit = NANOS_PER_SECOND / units_per_second as u32;
+
+    let (secs, subsec_units) = (value / units_per_second, value % units_per_second);
+    // See comment in copied Chrono code above for explanation of what's
+    // going on in this match statement.
+    //
+    // TODO[btv] - The expects below should never fail and are just here to document assumptions.
+    // Since they're potentially being called in a tight loop,
+    // we can optimize with `as` or unsafe code, if thisever proves to
```
Comment (umanwizard) [inline src/avro/src/decode.rs:199]:
```suggestion
    // we can optimize with `as` or unsafe code, if this ever proves to
```

**Comment by umanwizard** [comment]
I'm force-merging this because CI already passed on [7ce3ba2](https://github.com/MaterializeInc/materialize/pull/16609/commits/7ce3ba26c9819dc8a52c1a8b76078d2202bead06), and the only thing I've changed since then is comment text. CI now seems broken for unrelated reasons.

**Review by benesch — COMMENTED** [review]
Oops, forgot to press "submit" on this last night. Should we also upgrade chrono to the latest main, to get the bugfix in chrono itself?

File: src/avro/src/decode.rs [hunk 2]
```diff
@@ -110,22 +110,81 @@ impl Display for TsUnit {
...
+        //
+        // Thus, if there's somewhere else where we convert timestamps
+        // in seconds to Chrono `NaiveDateTime`s by calling
+        // `from_timestamp_opt(secs, 0)`, those timestamps (if negative)
+        // will fail to compare equal to ones generated by `build_ts_value`,
+        // even if they represent the same time.
+        //
+        // I think it's better if we just inherit Chrono's behavior for now here.
+        // I opened an issue with them to discuss:
+        // https://github.com/chronotope/chrono/issues/903
```
[the remark this answers is not in the document]
Comment (umanwizard) [reply src/avro/src/decode.rs]: I don't think we're using the old buggy `from_timestamp_millis` anywhere important, so we don't really need to upgrdae to the unreleased branch. I will check that, though.

[the remark this answers is not in the document]
Comment (benesch) [reply src/avro/src/decode.rs]: I think Petros is adding a use in https://github.com/MaterializeInc/materialize/pull/16622#pullrequestreview-1215694386.

[the remark this answers is not in the document]
Comment (umanwizard) [reply src/avro/src/decode.rs]: ok, let's look into upgrading Chrono in that case.

**Pull request merged**